fdb-extensions/src/main/java/com/apple/foundationdb/async/hnsw/HalfVector.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * HalfVector.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2025 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.async.hnsw;
+
+import com.apple.foundationdb.half.Half;
+
+import javax.annotation.Nonnull;
+import java.util.function.Supplier;
+
+/**
+ * A vector class encoding a vector over half components. Conversion to {@link DoubleVector} is supported and
+ * memoized.
+ */
+public class HalfVector extends AbstractVector {
+    @Nonnull
+    private final Supplier<DoubleVector> toDoubleVectorSupplier;
+
+    public HalfVector(@Nonnull final Half[] halfData) {
+        this(computeDoubleData(halfData));
+    }
+
+    public HalfVector(@Nonnull final double[] data) {
+        super(data);
+        this.toDoubleVectorSupplier = () -> new DoubleVector(data);
```

Comment:
I'm not sure, but part of me wonders if the `double`s here should be truncated to ensure that they fit in a fp16 value, so that serializing and deserializing a `HalfVector` is not lossy. For our purposes, maybe we don't really care about that